### PR TITLE
Adding in presenters per model.

### PR DESCRIPTION
**Pull request by cam156**
Includes adding in artifacts to refactor the submission presenter

I'm still unsure what to really call the ThinMint class.  It's a presenter factory Maybe?  Should it just be removed and use the helper for instantiating the AuthorPresenter?

This should be discussed before merging...

### CODE REVIEW

File: app/controllers/admin_controller.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def find_or_initialize_administrator
       end
       author = Author.find_or_initialize_by(access_id: Author.current.access_id)
       ### if in administrator list ????
-      authorize! :administer, :all
+      # authorize! :administer, :all
```
Comment (awead) [inline app/controllers/admin_controller.rb]: Did you mean to leave this commented-out?

File: app/helpers/presenter_helper.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module PresenterHelper
+  def current_author_presenter
+    presenter = @thin_mint.current_author_presenter if @thin_mint
+    presenter ||= AuthorPresenter.new(Author.current)
+    presenter
```
Comment (awead) [inline app/helpers/presenter_helper.rb]: I think you can remove this line.

File: app/helpers/presenter_helper.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  end
+
+  def model_presenter
+    presenter = @presenter
+    presenter ||= @thin_mint.model_presenter
+    presenter
```
Comment (awead) [inline app/helpers/presenter_helper.rb]: Same with this one.

File: app/models/submission_artifacts/artifact.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                :show_path
+
+    def initialize(submission)
+      @submission = submission
+      @states_can_be_updated = []
+      @states_can_be_created = []
+      @states_can_be_reviewed = []
+      @new_path = ''
+      @edit_path = ''
+      @show_path = ''
```
Comment (awead) [inline app/models/submission_artifacts/artifact.rb]: What's the reasoning behind setting these to empty? You could just check for nil instead.

File: app/models/submission_artifacts/artifact_generator.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+module SubmissionArtifacts
+  class ArtifactGenerator
+    def self.artifact_class_for_step(step_index)
+      artifact_list[step_index - 1]
```
Comment (awead) [inline app/models/submission_artifacts/artifact_generator.rb]: Is step_index ever 0?

File: app/models/submission_states/bogus.rb [hunk 6]
```diff
@@ -1,5 +1,9 @@
 module SubmissionStates
```
Comment (awead) [inline app/models/submission_states/bogus.rb]: If I was coming across this module for the first time, I'd be confused about what it's used for. Some YARD documentation would be nice.

File: app/models/submission_states/submission_state.rb [hunk 7]
```diff
@@ -9,6 +11,22 @@ def initialize
       @transitions_to = []
     end
 
+    def completion_date_method
+      ''
+    end
```
Comment (awead) [inline app/models/submission_states/submission_state.rb]: Are you doing checks for empty? Maybe nil instead? Or if the calling class is trying to determine if there is a method for determining the date, maybe `obj.respond_to?(:completion_date_method)`

File: app/models/submission_states/submission_state.rb [hunk 8]
```diff
@@ -9,6 +11,22 @@ def initialize
...
+      'completed'
+    end
+
+    def admin_state?
+      false
+    end
+
+    def rejection_date_method
+      ''
+    end
```
Comment (awead) [inline app/models/submission_states/submission_state.rb]: Same question as above.

File: app/presenters/committee_member_list_presenter.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
+class CommitteeMemberListPresenter
+  attr_reader :committee_members
+  attr_reader :submission
+
+  def initialize(submission)
+    @submission = submission
+    @committee_members = submission.committee_members.map { |committee_member| CommitteeMemberPresenter.new(committee_member) }
+  end
+
+  def committee_member_table(template)
```
Comment (awead) [inline app/presenters/committee_member_list_presenter.rb]: I'm not familiar with this design pattern of passing the view object in as a parameter. Did you see it used somewhere?

File: app/presenters/submission_presenter.rb [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+           :released_for_publication?,
+           to: :model
+
+  def initialize(model)
+    @model = model
+    @submission_state = SubmissionStates::StateGenerator.state_for_name(model.status)
+  end
+
+  def formatted_program_information
+    model.program_name + ' ' + model.degree.name + ' - ' + formatted_semester + ' ' + formatted_year
```
Comment (awead) [inline app/presenters/submission_presenter.rb]: Stylistic choice here, but this might be more readable:
`"#{model.program_name} #{model.degree.name} - #{formatted_semester} {#formatted_year}"`

File: app/views/author/submissions/_progress_indicator.html.erb [hunk 11]
```diff
@@ -1,55 +1,27 @@
...
-      <div class="status"><%= @submission_view.step_five_status %></div>
-    </div>
-    <div class="direction glyphicon glyphicon-arrow-down"></div>
-  </div>
-  <div class="step step-6 <%= @submission_view.step_six_class %>">
+  <%= render 'progress_indicator_step', submission_view: submission_view, step_index: 2 %>
+  <%= render 'progress_indicator_step', submission_view: submission_view, step_index: 3 %>
+  <%= render 'progress_indicator_step', submission_view: submission_view, step_index: 4 %>
+  <%= render 'progress_indicator_step', submission_view: submission_view, step_index: 5 %>
+  <div class="step step-6 <%= submission_view.class_for_step(6) %>">
```
Comment (awead) [inline app/views/author/submissions/_progress_indicator.html.erb]: Could you get this to work like: `<%= render 'progress_indicator_step', submission_view: submission_view, step_index: 6 %>`? You could then have a step-agnostic method for arrow.

File: app/views/author/submissions/_progress_indicator.html.erb [hunk 12]
```diff
@@ -1,55 +1,27 @@
...
     <div class="information">
       <div class="description"><%= I18n.t("#{current_partner.id}.partner.name") %> approves Final Submission files</div>
-      <div class="status"><%= @submission_view.step_six_status %></div>
+      <div class="status"><%= submission_view.step_status(self, 6) %></div>
     </div>
-    <%= @submission_view.step_seven_arrow %>
+    <%= submission_view.step_seven_arrow(self) %>
   </div>
-  <div class="step step-7"<%= @submission_view.step_seven_class %>">
+  <div class="step step-7"<%= submission_view.class_for_step(7) %>">
```
Comment (awead) [inline app/views/author/submissions/_progress_indicator.html.erb]: Same here: `<%= render 'progress_indicator_step', submission_view: submission_view, step_index: 7 %>` with maybe  some minimal logic in `_progress_indicator_step.html.erb` or additional presenter/helper methods?

File: app/views/author/submissions/_progress_review_link.html.erb [hunk 13]
```diff
@@ -0,0 +1 @@
+<%= link_to('[review]', url, class: 'small') %>
```
Comment (awead) [inline app/views/author/submissions/_progress_review_link.html.erb]: This might be a good use of a helper instead of a one-line file, but I might be missing something.

File: app/views/author/submissions/_progress_update_link.html.erb [hunk 14]
```diff
@@ -0,0 +1 @@
+<%= link_to('[update]', url, class: 'small') %>
```
Comment (awead) [inline app/views/author/submissions/_progress_update_link.html.erb]: Same comment as above.

File: spec/component/models/submission_artifacts/committee_spec.rb [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+                   reviewable?:    true,
+                   updatable?:     false,
+                   creatable?:     false,
+                   creation_step?: false,
+                   rejected?:      false,
+                   created?:       true)
+    end
+  end
+
+  def test_methods(artifact, expected_results)
```
Comment (awead) [inline spec/component/models/submission_artifacts/committee_spec.rb]: This method is repeated three other times in the subsequent tests. It looks the same, so maybe it can be extracted to a helper file and loaded so it's shareable.

File: spec/integration/author/submission_status_giver_spec.rb [hunk 16]
```diff
@@ -347,6 +347,7 @@
     context "visiting the 'Review Committee' page" do
       before { visit author_submission_committee_members_path(submission) }
       specify "displays the committee review page" do
+        puts page.text
```
Comment (awead) [inline spec/integration/author/submission_status_giver_spec.rb]: remove?

File: spec/views/view_spec_helper.rb [hunk 17]
```diff
@@ -1,3 +1,14 @@
 # require 'rails_helper'
 require 'spec_helper'
 require 'support/rails'
+require 'support/factories'
+require 'support/database_cleaner'
+require 'support/committee_factory'
+# require 'support/keyword_factory'
```
Comment (awead) [inline spec/views/view_spec_helper.rb]: remove?

**Comment by awead** [comment]
@cam156 :clap: yay, refactoring! The only concern I have is how we collectively think about what presenters actually are. For example, you seem to be using presenter methods like `committee_member_table` to generate html, which in my mind is the function of a view. When I think of presenters, I think of objects that deal with logic and manipulation of data that is in models, but I don't think of them as actually responsible for generating html. This is a broader question that would impact how we want to use these classes moving forward, and it isn't meant to prevent this PR from being merged.

I'm only bringing this up now because I probably would have implemented these very differently than you had I been tasked with it. This isn't a right way versus wrong way kind of thing, it's about agreement so we can all be consistent in the future

**Comment by cam156** [comment]
@awead thanks for the review.  You raise a good question,  Should all the html be moved into helpers or partials instead of being inside a presenter?  I obviously leaned towards it being in the presenter and the previous iteration had html hard coded in strings in the presenter.  That said I'm ok moving it.

File: spec/component/models/submission_artifacts/committee_spec.rb (hunk 15 above)
Comment (cam156) [inline spec/component/models/submission_artifacts/committee_spec.rb]: @awead :+1:

File: app/helpers/presenter_helper.rb (hunk 3 above)
Comment (cam156) [inline app/helpers/presenter_helper.rb]: @awead rubocop was not happy with that syntax

File: app/controllers/admin_controller.rb (hunk 1 above)
Comment (cam156) [inline app/controllers/admin_controller.rb]: It should likely be in another commit, but I did intend to remove it.  We are already authorized at this point.

File: app/helpers/presenter_helper.rb (hunk 2 above)
Comment (cam156) [inline app/helpers/presenter_helper.rb]: Same as below.  This line made rubocop happy.

File: app/models/submission_artifacts/artifact.rb (hunk 4 above)
Comment (cam156) [inline app/models/submission_artifacts/artifact.rb]: In this case that might be true.  I was trying to avoid having to nil check everywhere.  Blank is my bogus path.

File: app/models/submission_artifacts/artifact_generator.rb (hunk 5 above)
Comment (cam156) [inline app/models/submission_artifacts/artifact_generator.rb]: The steps in the UI are 1through 7.  I was trying to minimize how much I changed by sticking with the original classes eat.

File: app/presenters/committee_member_list_presenter.rb (hunk 9 above)
Comment (cam156) [inline app/presenters/committee_member_list_presenter.rb]: @awead I came across it when researching presenters.  We could also include the view helpers into the presenter.  This gives a much more specific context, but I am not married to it.

File: app/views/author/submissions/_progress_indicator.html.erb (hunk 11 above)
Comment (cam156) [inline app/views/author/submissions/_progress_indicator.html.erb]: @awead. I had gone through so much refactoring to get to this point.  :+1: to this, but maybe as another step.

**Comment by awead** [comment]
@cam156 Since you've already done the work here, I don't want to you to redo it again :smile: so I think the approach is sound for this PR. I think it's something to mull over later, though. My inclination is:

1. No html in presenters, only in views and helpers
2. Presenters are manipulating model data for a view, i.e. it gets model-based logic
3. Helpers are for generic html manipulation and simpler logic, not tied to a specific model

Again, just some initial thoughts. Although I am more heavily inclined to the first point than the others, I'm open to different ideas.

File: app/views/author/submissions/_progress_indicator.html.erb (hunk 11 above)
Comment (awead) [inline app/views/author/submissions/_progress_indicator.html.erb]: @cam156 yes, you can definitely call this done for now.

File: app/models/submission_artifacts/artifact_generator.rb (hunk 5 above)
Comment (awead) [inline app/models/submission_artifacts/artifact_generator.rb]: :ok:

File: app/models/submission_artifacts/artifact.rb (hunk 4 above)
Comment (cam156) [inline app/models/submission_artifacts/artifact.rb]: @awead Just double checked.  I just use the string as is in the code.  No reason for a nil check if I set the strings to blank.

File: app/models/submission_states/submission_state.rb (hunk 7 above)
Comment (cam156) [inline app/models/submission_states/submission_state.rb]: @awead I am just using the name as is, so no checking for nil.  Blank just "works".  Although I say works in quotes becuase if you send nil or empty string to an object they return a method_missing error.

@awead The actual method changed based on the state the submission is in, if you are uploading format_review_files then the completion date is "format_review_files_uploaded_at" on the submission.  If you are uploading final files then the  date is "final_submission_files_uploaded_at" is you are waiting for a review of your files the date of completion is yet a different date.  

This method definition is just to be certain that all submission states have a completion date even though not every one does.

I have either named this wrong or gone about it wrong if it needs this much explanation.  Any ideas on making it more clear?

File: app/models/submission_states/submission_state.rb (hunk 8 above)
Comment (cam156) [inline app/models/submission_states/submission_state.rb]: @awead similar answer to above.  Since the state does not have a record of the submission I was just keeping the method name.

I wonder if it would be better to have the public method rejection_date( submission) which returned the actual date?

I'm up for suggestions on how to make this concept more clear.

File: app/presenters/submission_presenter.rb (hunk 10 above)
Comment (cam156) [inline app/presenters/submission_presenter.rb]: :ok_hand:

File: app/views/author/submissions/_progress_review_link.html.erb (hunk 13 above)
Comment (cam156) [inline app/views/author/submissions/_progress_review_link.html.erb]: @awead and others  this brings up a good question on when to use helpers vs partials.  This code is only used in one view.  The rule of thumb I had read was helpers should be universally used.  That said I'm willing for us to come up with our own opinion.

This question goes along with Adam's overall question about where the HTML should go.  Do we extract all the html from the presenter?  If we do, do we generate lots of little helper methods, or lots of little partials for those html fragments?  These two partials were originally part of the presenter as I refactored, but since they were "plain" html with no logic I moved them into a partial.  That begs the question of why I did not move all the other html out.  The answer was quite simply that the refactoring path I took created these guys as little methods without logic and the other html was still surrounded by logic in the presenter.  If this then show this...

File: app/controllers/admin/authors_controller.rb [hunk 18]
```diff
@@ -1,19 +1,19 @@
...
 
   def edit
-    @author = Author.find(params[:id])
+    @presenter = AuthorPresenter.new(Author.find(params[:id]))
   end
 
   def update
-    @author = Author.find(params[:id])
-    @author.update_attributes!(author_params)
+    author = Author.find(params[:id])
```
Comment (hectorcorrea) [inline app/controllers/admin/authors_controller.rb]: +1 for preferring local variables

File: app/presenters/thin_mint.rb [hunk 19]
```diff
@@ -0,0 +1,18 @@
+class ThinMint
```
Comment (hectorcorrea) [inline app/presenters/thin_mint.rb]: Maybe a description why this is called ThinMint would be helpful in the future.

File: app/controllers/static_controller.rb [hunk 20]
```diff
@@ -1,4 +1,4 @@
-class StaticController < AuthorController
+class StaticController < ApplicationController
```
Comment (hectorcorrea) [inline app/controllers/static_controller.rb]: Wow....I can't believe this controller was inheriting from AuthorController.

File: app/models/submission_artifacts/artifact.rb [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    # has the artifact already been created at which point we can only update
+    def created?
+      false
```
Comment (hectorcorrea) [inline app/models/submission_artifacts/artifact.rb]: How come these are hard-coded?

Comment (hectorcorrea) [inline app/models/submission_artifacts/artifact.rb]: Never mind...I just saw the code in the sub-classes.

File: app/models/submission_states/submission_state.rb (hunk 7 above)
Comment (hectorcorrea) [inline app/models/submission_states/submission_state.rb]: What trip me on this method is the word "date". After looking at the other instances of this method it is clear it does not have anything to do with dates. Perhaps adding a comment in this case will be a good idea. 

I am not sure if using something like `:bogus` instead of the empty string will be clearer (and I am not sure what such change will do to the logic)

File: app/presenters/author_list_presenter.rb [hunk 22]
```diff
@@ -0,0 +1,13 @@
+class AuthorListPresenter
+  attr_reader :authors
+
+  def initialize(authors)
+    @authors = authors.map { |author| AuthorPresenter.new(author) }
+  end
+
+  def to_rabl(template)
```
Comment (hectorcorrea) [inline app/presenters/author_list_presenter.rb]: I didn't know what rabl was, is it this? https://github.com/nesquena/rabl

File: app/presenters/committee_member_list_presenter.rb (hunk 9 above)
Comment (hectorcorrea) [inline app/presenters/committee_member_list_presenter.rb]: Not knowing much about presenters and how they are created I appreciated seeing the parameter on the method since I didn't have to guess what template referred to.

File: app/presenters/submission_presenter.rb [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+    template.content_tag(:span, class: 'delete-link') do
+      template.link_to('[delete]', template.author_submission_path(model.id), class: 'text-danger', method: 'delete', confirm: 'Permanently delete this submission?', rel: 'nofollow')
+    end
+  end
+
+  def created_on
+    model.created_at.strftime('%B %-e, %Y')
+  end
+
+  def step_seven_status(_template)
```
Comment (hectorcorrea) [inline app/presenters/submission_presenter.rb]: Maybe this should be called `last_step_status` ? It sort of bugs me seeing `seven` on the method name but maybe it does not matter.

File: app/presenters/submission_presenter.rb [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+      'complete'
+    else
+      ''
+    end
+  end
+
+  def step_status(template, step_index)
+    step_state = state_from_step(step_index)
+    if artifact_for_step(step_index).rejected?
+      ("#{template.error_icon} rejected#{formatted_timestamp_of(model.send(step_state.rejection_date_method))}, please see the <a href='" + "/author/submissions/#{id}/final_submission/edit#final-submission-notes" + "'>notes from the administrator</a>").html_safe
```
Comment (hectorcorrea) [inline app/presenters/submission_presenter.rb]: I would break this line into smaller tokens to make it easier to read.

File: app/presenters/submission_presenter.rb [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+      completed_step(template, step_index, step_state)
+    end
+  end
+
+  private
+
+    def completed_step(template, step_index, step_state = nil)
+      return '' unless current_is_beyond_step?(step_index)
+      step_state ||= state_from_step(step_index)
+      "#{template.ok_icon} #{step_state.completion_word}#{formatted_timestamp_of(model.send(step_state.completion_date_method))}".html_safe
```
Comment (hectorcorrea) [inline app/presenters/submission_presenter.rb]: Same here, I would extract the `formatted_timestamp_of(model.send(step_state.completion_date_method))` out of this line into something like this:

```
date_formatted = formatted_timestamp_of(model.send(step_state.completion_date_method))
"#{template.ok_icon} #{step_state.completion_word}#{date_formated}".html_safe
```

File: app/views/author/committee_members/_committee_member_fields.html.erb [hunk 26]
```diff
@@ -1,26 +1,3 @@
...
-  <%= f.input :is_required, as: :hidden, :input_html => {value: f.object.is_required, readonly: true } %>
-  <div class="row">
-    <div class="col-sm-6">
-        <%= f.input :name, input_html: { class: 'ldap-lookup' } %>
-    </div>
-    <div class="col-sm-6">
-        <%= f.input :email %>
-    </div>
-  </div>
+  <%== lpresenter.show_member(f, self) %>
```
Comment (hectorcorrea) [inline app/views/author/committee_members/_committee_member_fields.html.erb]: Sweeeeeeet!

Comment (hectorcorrea) [inline app/views/author/committee_members/_committee_member_fields.html.erb]: First I thought this was sweet: 24 lines removed. Then I realized that the structure of the HTML generated is now lost.
I thought the goal of the presenter was to remove logic but in this case we are removing both logic and the structure of the HTML. Mixed feelings.

File: app/views/author/submissions/_progress_indicator.html.erb [hunk 27]
```diff
@@ -1,55 +1,27 @@
...
-  <div class="step step-5 <%= @submission_view.step_five_class %>">
-    <div class="number">5</div>
-    <div class="information">
-      <div class="description"><%= @submission_view.step_five_description %></div>
-      <div class="status"><%= @submission_view.step_five_status %></div>
-    </div>
-    <div class="direction glyphicon glyphicon-arrow-down"></div>
-  </div>
-  <div class="step step-6 <%= @submission_view.step_six_class %>">
+  <%= render 'progress_indicator_step', submission_view: submission_view, step_index: 2 %>
```
Comment (hectorcorrea) [inline app/views/author/submissions/_progress_indicator.html.erb]: This is much easier to read than the previous version!!!!

File: app/views/author/submissions/_progress_review_link.html.erb (hunk 13 above)
Comment (hectorcorrea) [inline app/views/author/submissions/_progress_review_link.html.erb]: Hard to come with a one-size-fits-all approach. I like when the HTML shows the structure of the view to be rendered and has little or no logic on it. But as you mentioned at times this results into HTML in the presenter or little tiny partials. Not sure I have an answer for this.

File: spec/factories/submissions.rb [hunk 28]
```diff
@@ -37,6 +41,7 @@
     trait :waiting_for_final_submission_response do
       status "waiting for final submission response"
       format_review_notes "Format review notes"
+      final_submission_files_uploaded_at Time.zone.local(2014, 9, 1, 0, 0, 0)
```
Comment (hectorcorrea) [inline spec/factories/submissions.rb]: Where did this date came from?

**Comment by hectorcorrea** [comment]
Awesome job Carolyn. I left a few comments here and there but I am OK if you just want to merge as-is.

File: app/presenters/author_list_presenter.rb (hunk 22 above)
Comment (jonibar) [inline app/presenters/author_list_presenter.rb]: Yes that is rabl.  The rabl templates are used on the submission index pages.  The rabl templates format submission data for the javascript datatables.

File: app/models/submission_states/submission_state.rb (hunk 7 above)
Comment (cam156) [inline app/models/submission_states/submission_state.rb]: @hector It does have to do with a Date, or I guess a DateTime.  This is the name of the method to get the timestamp for when the submission state was completed.  The _at in the other methods implies a date/time.

Would completed_at_method be a better name?

File: app/presenters/submission_presenter.rb (hunk 23 above)
Comment (cam156) [inline app/presenters/submission_presenter.rb]: @hectorcorrea that was copied from the original code from the view level presenter.  This PR was so big I had to stop some where so we could start discussing the concept.

File: app/presenters/submission_presenter.rb (hunk 24 above)
Comment (cam156) [inline app/presenters/submission_presenter.rb]: @hectorcorrea again a straight copy.  It is a very fine line between getting something out to discuss and move forward with and fixing every place where html is hard coded.

File: app/views/author/committee_members/_committee_member_fields.html.erb (hunk 26 above)
Comment (cam156) [inline app/views/author/committee_members/_committee_member_fields.html.erb]: @hectorcorrea Yeah...  I was trying to remove the if else from the code.  Likely a better way to do this.  This also goes to @awead's question about html in the presenter.
